### ISSUE TITLE
init keyword is clunky

### ISSUE THREAD

**Issue by theangrydev** [JetBrains/spek#41]
I don't like having to do { init { } } instead of {{ }}, have you guys any ideas about how you might be able to land on a clean syntax again?

**Comment by paslavsky** [comment]
I found one solution, maybe it help
Please see my [pull request](https://github.com/JetBrains/spek/pull/48)

**Comment by hastebrot** [comment]
Amazing! So simple. We still need parenthesis `()` around the inline function, but this is much better than `init {}`.

Before:
~~~kotlin
class CalculatorConsoleSpecs: org.jetbrains.spek.api.Spek() {
    init {
    given("a calculator") {
        // ...
    }
}
}
~~~

After:
~~~kotlin
class CalculatorConsoleSpecs: org.jetbrains.spek.api.Spek({
    given("a calculator") {
        // ...
    }
})
~~~

**Comment by mhshams** [comment]
I think, a more practical solution would be to move the scenarios inside methods. 

```kotlin
class MySpek {
   
   @Scenario("title of the fist scenario")
    fun firstScenario() {
            given("a calculator") {
                  on("...") {
                      then("...") {
                      }
            } 
    }
    
   @Scenario("title of the second scenario")
    fun secondScenario() {
            given("a calculator") {
                  on("...") {
                      then("...") {
                      }
            } 
    }
}
```
Spek runner can scan the classes for the specific annotation (e.g. Scenario). 

Advantages: 
 - all scenarios that are related to a single feature would be in a single class. 
 - no need to worry about 'init' any more

**Comment by hastebrot** [comment]
@mhshams The issue here is, that we need to find a method name (e.g. `firstScenario()`), which is redundant. 

The Spock testing framework for Groovy solved this, by accepting strings as method names. Also some JavaScript and Ruby testing frameworks avoid this, by _calling_ functions that accept strings for BDD declarations (e.g. `setup()`, `describe()`, `it()`), instead of _declaring_ functions.

**Comment by mhshams** [comment]
@hastebrot , long time ago, when I was playing with spek, I suggested to use the **kotlin escape character** to define the method names as string. 
something like: 
```kotlin
class MySpek {

    fun `title of the fist scenario`() {
            given("a calculator") {
                  on("...") {
                      then("...") {
                      }
            } 
    }

    fun `title of the second scenario`() {
            given("a calculator") {
                  on("...") {
                      then("...") {
                      }
            } 
    }
}
```
With this approach you don't need have the annotation or another top declaration.

**Comment by hastebrot** [comment]
@mhshams Wow, thanks for this! I didn't know this was possible in Kotlin. Even IDEA supports auto-completion for calling these methods with escaped strings.

**Comment by hhariri** [comment]
@mhshams I like that approach in the benefits it offers. I'm just not such a fan of the escape characters. However since we first discussed it, things have also changed in Kotlin, including the need for init, so this might be ultimately a better approach. 

In addition, it will solve some of the issues we have. Wondering if beforeOn/afterOn would be solved too...Going to play around with this. 

(Btw, your image update, on my list, but given we need to change things anyway, I've not updated yet)

@paslavsky Thanks for the PR. Also going to have a play.

Btw, on our Slack channel, there's a Spek group. I've pointed people here, but maybe you'd like to join the discussions there too? 

Sign up here: http://kotlinslackin.herokuapp.com
